geps/gep-1713/index.md

```diff
@@ -0,0 +1,377 @@
+	// Group is the group of the referent. For example, "gateway.networking.k8s.io".
+	// When unspecified or empty string, core API group is inferred.
+	//
+	// +optional
+	// +kubebuilder:default=""
+	Group *Group `json:"group"`
+
+	// Kind is kind of the referent. For example "Gateway".
+	//
+	// +optional
+	// +kubebuilder:default=Gateway
+	Kind *Kind `json:"kind"`
+
+	// Name is the name of the referent.
+	Name ObjectName `json:"name"`
+
+	// Namespace is the namespace of the referenced object. When unspecified, the local
+	// namespace is inferred.
+	//
+	// Support: Core
+	//
+	// +optional
+	Namespace *Namespace `json:"namespace,omitempty"`
+}
+```
+
+#### YAML
+
+Below shows an example of an end to end configuration with Gateway merging.
+Here we define a parent resource, which allows children from the same namespace.
+A single listener is defined in the parent.
+
+The child gateway attaches to this Gateway and specifies an additional listener.
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: parent-gateway
+spec:
+  gatewayClassName: example
+  infrastructure:
+    allowedChildren:
+      namespaces:
+        from: Same
+  listeners:
+  - name: common-monitoring
+    port: 8080
+    protocol: HTTP
+---
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: child-gateway
+spec:
+  gatewayClassName: example
+  infrastructure:
+    attachTo:
+      name: parent-gateway
+      kind: Gateway
+      group: gateway.networking.k8s.io
+  listeners:
+  - name: domain-a
+    hostname: a.example.com
+    protocol: HTTP
+    port: 80
+```
+
+Logically, this is equivalent to a single Gateway as below:
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: gateway
+spec:
+  gatewayClassName: example
+  listeners:
+  - name: common-monitoring
+    port: 8080
+    protocol: HTTP
+  - name: domain-a
+    hostname: a.example.com
+    protocol: HTTP
+    port: 80
+```
+
+### Semantics
+
+#### Route Attaching
+
+Routes MUST be able to specify a "child" Gateways as a `parentRef` and make use of the fields in `ParentReference` to help target a specific listener.
+
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: HTTPRoute
+metadata:
+  name: httproute-example
+spec:
+  parentRefs:
+  - name: child-gateway
+    sectionName: metrics
+```
+
+Routes can only bind to listeners *directly* defined in the `Gateway` referenced.
+For instance, the following route referencing a listener defined in the parent `Gateway`, but attaching to the child `Gateway` is not valid.
+This will be reported in [status](#status-fields).
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: HTTPRoute
+metadata:
+  name: httproute-example
+spec:
+  parentRefs:
+  - name: child-gateway
+    sectionName: section-in-parent
+```
+
+#### Merging Spec Fields
+
+##### Validation
+
+A parent and child MUST have the same `gatewayClassName`.
```

Comment:
Is this required or can the child simply omit. Omission is likely better if the root gateway is infrastructure that is being managed centrally. The children don't actually need to know or care about the class.

Reply:
I think it would great to not require it.

I'm not familiar if it's possible with CEL to make a field optional if some other property is set (eg. infrastructure.attachTo`

@robscott do you know - I'd consider you the CEL expert.

Reply:
You can do it, you just need to attach the rule to some common root (here that would be `spec`), rather than on `gatewayClassName`

Reply:
It does make sense that the class needs to be managed centrally (by the parent Gateway administrator), but`spec.gatewayClassName` is a required field on [Gateway](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1alpha2.Gateway) and while the field doesn't appear to actually be immutable, it's [recommended that the specified GatewayClass be used as a template](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1alpha2.GatewayClass) when creating a Gateway so while it's a bit of extra boilerplate I don't expect it's something that should ever be changing dynamically.

What we may want to add is a clarification that about the GatewayClass "template snapshotting" recommendation - this should likely be tied to the parent Gateway rather than attempting to reconcile potentially drifting GatewayClass configuration snapshots from multiple child Gateways.

Reply:
While I get the templating recommendation in principle I think it presents a toil problem for users when you attempt to apply it to the specification of Gateway.gatewayClassName.  

Since the compositional model for routes & policies is back-reference the toil fanout of forcing Gateway immutability is high on users. In this situation any class transition (infrastructure, version etc.) would force a reference update on every inbound policy since the user would have to create a new Gateway resource. 

Even immutability is somewhat of an illusion since the recommendation can't prevent the deletion & recreation of the resource with the same name and a difference value. The user is simply left with the side-effects of their workaround.

Reply:
For clarity, I would be in favor of allowing an empty `gatewayClassName` contingent on `attachTo` being configured if that's permissible - I just think it's a nicety rather than a necessity because changing a `gatewayClassName` will often be a destructive action akin to deleting and recreating all child and parent resources.

Reply:
Again, `gatewayClassName` is a required field, and changing it to optional is a breaking API change.

Reply:
I don't think we can remove `gatewayClassName` because if we try to attach to a non-existent parent then who would be responsible to update the status of the Gateway saying invalid references etc?

I'm going to resolve this thread end of week unless someone has an idea on solving the situation I just mentioned.

Reply:
> I don't think we can remove gatewayClassName because if we try to attach to a non-existent parent then who would be responsible to update the status of the Gateway saying invalid references etc?

That makes a lot of sense to me, hadn't considered that situation.